File: genetic_algorithm/genetic_algorithm-prolog/genetic_algorithm.pl
:- use_module(library(http/json)).
:- initialization(main).

:- dynamic individuo/7. % espacos, valores, cromossomo, espaco_usado, limite_espacos, geracao, nota_avaliacao

% Le um arquivo JSON
lerJson(File) :-
 open("../produtos.json", read, Teste),
 json_read_dict(Teste, File).

% Retorna um intervalo de uma lista
sliceArray([], _, _, [], _).
sliceArray([H|T], Start, End, [H|Tail], Cont) :- 
 Cont >= Start, Cont < End, NewCont is Cont+1, sliceArray(T, Start, End, Tail, NewCont).
sliceArray([_|T], Start, End, Elem, Cont) :- NewCont is Cont+1, sliceArray(T, Start, End, Elem, NewCont).

% Concatena duas Listas
concatArray(L1, L2, L3) :- append(L1, L2, L3).

% Retorna elemento pelo indice na lista
getByIndex([], _, _) :- fail.
getByIndex([H|_], 0, H).
getByIndex([_|T], Index, E) :- Index > 0, NewIndex is Index-1, getByIndex(T, NewIndex, E).

% Float (Scientific Notation) number to String
float_to_string(Number, String, Size) :- format(atom(A), Size, [Number]), atom_string(A, String).

% Retorna os produtos
getProdutos(Produtos) :- lerJson(Produtos).

getEspacos([], []).
getEspacos([P|T], [H|Tail]) :- H = P.espaco, getEspacos(T, Tail).

getValores([], []).
getValores([P|T], [H|Tail]) :- H = P.valor, getValores(T, Tail).

getNomes([], []).
getNomes([P|T], [H|Tail]) :- H = P.nome, getNomes(T, Tail).

% Retorna o tamanho da maior string de nome
getTamanhoStringProdutoComMaiorNome([], Maior, Maior).
getTamanhoStringProdutoComMaiorNome([H|T], Aux, Maior) :- 
 string_length(H.nome, StrSize), StrSize > Aux, getTamanhoStringProdutoComMaiorNome(T, StrSize, Maior).
getTamanhoStringProdutoComMaiorNome([_|T], Aux, Maior) :- getTamanhoStringProdutoComMaiorNome(T, Aux, Maior).

% Retorna o tamanho da maior string de espaco
getTamanhoStringProdutoComMaiorEspaco([], Maior, Maior).
getTamanhoStringProdutoComMaiorEspaco([H|T], Aux, Maior) :- float_to_string(H.espaco, Str, "~8f"),
 string_length(Str, StrSize), StrSize > Aux, getTamanhoStringProdutoComMaiorEspaco(T, StrSize, Maior).
getTamanhoStringProdutoComMaiorEspaco([_|T], Aux, Maior) :- getTamanhoStringProdutoComMaiorEspaco(T, Aux, Maior).

% Retorna o tamanho da maior string de valor
getTamanhoStringProdutoComMaiorValor([], Maior, Maior).
getTamanhoStringProdutoComMaiorValor([H|T], Aux, Maior) :- float_to_string(H.valor, Str, "~8f"),
 string_length(Str, StrSize), StrSize > Aux, getTamanhoStringProdutoComMaiorValor(T, StrSize, Maior).
getTamanhoStringProdutoComMaiorValor([_|T], Aux, Maior) :- getTamanhoStringProdutoComMaiorValor(T, Aux, Maior).

% Rretorna as infos necessarias para imprimir os produtos
getMaioresTamanhosStringProdutosInfo(Produtos, [MN, ME, MV]) :- 
 getTamanhoStringProdutoComMaiorValor(Produtos, 0, MV), getTamanhoStringProdutoComMaiorEspaco(Produtos, 0, ME), getTamanhoStringProdutoComMaiorNome(Produtos, 0, MN).

% Imprime x espacos na mesma linha
imprimeEspacos(0) :- write('').
imprimeEspacos(X) :- write(' '), NewX is X-1, imprimeEspacos(NewX).

% Imprime os produtos
imprimeProdutosAux([], _).
imprimeProdutosAux([H|T], [MN,ME,MV]) :-
 write(H.nome), string_length(H.nome, NL), NewMN is MN - NL, imprimeEspacos(NewMN), write(" - "),
 float_to_string(H.valor, Str1, "~5f"), write(Str1),  string_length(Str1, VL), NewMV is MV - VL, imprimeEspacos(NewMV), write(" - "),
 float_to_string(H.espaco, Str2, "~8f"), write(Str2), write_ln(''), imprimeProdutosAux(T, [MN,ME,MV]).

imprimeProdutos(Produtos) :- getMaioresTamanhosStringProdutosInfo(Produtos, [MN,ME,MV]),
 write("Produto"), NewMN is MN-6, imprimeEspacos(NewMN), 
 write("Valor"), NewMV is MV-2, imprimeEspacos(NewMV), 
 write_ln("Espaco"), imprimeProdutosAux(Produtos, [MN,ME,MV]).

% Cria um individuo
cromossomo([], 0).
cromossomo([H|T], S) :- NS is S-1, random(0,2,H), cromossomo(T, NS).

initIndividuo(Espacos, Valores, Limite, Individuo) :- length(Espacos, Size), cromossomo(C, Size),
 Individuo = individuo(Espacos, Valores, C, 0, Limite, 0, 0).

% Visualiza um individuo
visualizaIndividuo(Individuo) :-
 getIndividuoGeracao(Individuo, G), write("Geracao: "), write_ln(G),
 getIndividuoNotaAvaliacao(Individuo, N), write("Nota avaliacao R$: "), write_ln(N),
 getIndividuoEspacoUsado(Individuo, E), write("Espaco usado: "), write_ln(E),
 getIndividuoCromossomo(Individuo, C), write("Cromossomo: "), write_ln(C), write_ln("").

% Gets para o individuo
getIndividuoEspacos(individuo(Espacos, _, _, _, _, _, _), Espacos).
getIndividuoValores(individuo(_, Valores, _, _, _, _, _), Valores).
getIndividuoCromossomo(individuo(_, _, Cromossomo, _, _, _, _), Cromossomo).
getIndividuoEspacoUsado(individuo(_, _, _, EspacoUsado, _, _, _), EspacoUsado).
getIndividuoLimiteEspaco(individuo(_, _, _, _, LimiteEspaco, _, _), LimiteEspaco).
getIndividuoGeracao(individuo(_, _, _, _, _, Geracao, _), Geracao).
getIndividuoNotaAvaliacao(individuo(_, _, _, _, _, _, NotaAvaliacao), NotaAvaliacao).

% Avalia o cromossomo de um individuo
avaliacaoCromossomoIndividuo(Individuo, _, _, SomaEspaco, _, (1, SomaEspaco)) :- getIndividuoLimiteEspaco(Individuo, LE), SomaEspaco >= LE.
avaliacaoCromossomoIndividuo(_, [], Nota, SomaEspaco, _, (Nota, SomaEspaco)).
avaliacaoCromossomoIndividuo(Individuo, [H|T], Nota, SomaEspaco, Index, Return) :- H =:= 1, 
 getIndividuoValores(Individuo, Valores), getByIndex(Valores, Index, V), NewNota is Nota + V,
 getIndividuoEspacos(Individuo, Espacos), getByIndex(Espacos, Index, E), NewSomaEspaco is SomaEspaco + E,
 NewIndex is Index+1, avaliacaoCromossomoIndividuo(Individuo, T, NewNota, NewSomaEspaco, NewIndex, Return).
avaliacaoCromossomoIndividuo(Individuo, [_|T], Nota, SomaEspaco, Index, Return) :- NewIndex is Index+1,
 avaliacaoCromossomoIndividuo(Individuo, T, Nota, SomaEspaco, NewIndex, Return).

% Avalia um individuo
avaliacaoIndividuo(Individuo, Return) :-
 getIndividuoEspacos(Individuo, E), getIndividuoValores(Individuo, V), getIndividuoCromossomo(Individuo, C),
 getIndividuoLimiteEspaco(Individuo, LE), getIndividuoGeracao(Individuo, G), 
 avaliacaoCromossomoIndividuo(Individuo, C, 0, 0, 0, (NewNotaAvaliacao, NewEspacoUsado)),
 Return = individuo(E, V, C, NewEspacoUsado, LE, G, NewNotaAvaliacao).  

% Faz o crossover de individuos
corte(Size, Return) :- random(1, Size, Return).

crossoverIndividuo(A, B, (Filho1, Filho2)) :-
 getIndividuoCromossomo(A, C1), getIndividuoCromossomo(B, C2), length(C1, Size1), length(C2, Size2),
 getIndividuoGeracao(A, G), NewG is G+1, getIndividuoValores(A, V), getIndividuoEspacos(A, E),
 getIndividuoLimiteEspaco(A, LE),

 corte(Size1, Corte),

 sliceArray(C2, 0, Corte, Filho1_pt1, 0),
 sliceArray(C1, Corte, Size1, Filho1_pt2, 0),
 concatArray(Filho1_pt1, Filho1_pt2, F1),

 sliceArray(C1, 0, Corte, Filho2_pt1, 0),
 sliceArray(C2, Corte, Size2, Filho2_pt2, 0),
 concatArray(Filho2_pt1, Filho2_pt2, F2),

 avaliacaoCromossomoIndividuo(A, F1, 0, 0, 0, (NewNotaAvaliacao1, NewEspacoUsado1)),
 avaliacaoCromossomoIndividuo(A, F2, 0, 0, 0, (NewNotaAvaliacao2, NewEspacoUsado2)),

 Filho1 = individuo(E, V, F1, NewEspacoUsado1, LE, NewG, NewNotaAvaliacao1),
 Filho2 = individuo(E, V, F2, NewEspacoUsado2, LE, NewG, NewNotaAvaliacao2).

% Faz mutacao no cromossomo de um individuo
getCromossomoMutacaoAux(N, Taxa, Random, 0) :- Random < Taxa, N =:= 1.
getCromossomoMutacaoAux(N, Taxa, Random, 1) :- Random < Taxa, N =:= 0.
getCromossomoMutacaoAux(N, _, _, N).

getCromossomoMutacao([], _, []).
getCromossomoMutacao([H|T], Taxa, [N|Tail]) :- random(Rand), getCromossomoMutacaoAux(H, Taxa, Rand, N),
 getCromossomoMutacao(T, Taxa, Tail).

mutacaoIndividuo(Individuo, Taxa, NewIndividuo) :-
 getIndividuoCromossomo(Individuo, C), getIndividuoGeracao(Individuo, G),
 getIndividuoValores(Individuo, V), getIndividuoEspacos(Individuo, E),
 getIndividuoLimiteEspaco(Individuo, LE), getIndividuoEspacoUsado(Individuo, EU),
 getIndividuoNotaAvaliacao(Individuo, NA),

 getCromossomoMutacao(C, Taxa, NewCromossomo),
 NewIndividuo = individuo(E, V, NewCromossomo, EU, LE, G, NA).

% Cria a populacao
initPopulacao(0, _, _, _, []).
initPopulacao(Tamanho, Espacos, Valores, LimiteEspaco, [H|T]) :-
 initIndividuo(Espacos, Valores, LimiteEspaco, H),
 NewTamanho is Tamanho-1,
 initPopulacao(NewTamanho, Espacos, Valores, LimiteEspaco, T).

% Retorna o melhor individuo
melhorIndividuo(I1, I2, I1) :- getIndividuoNotaAvaliacao(I1, N1), getIndividuoNotaAvaliacao(I2, N2), N1 >= N2.
melhorIndividuo(_, I2, I2).

% Ordena a populacao
getPairs([], []).
getPairs([H|T], [NA-H|Tail]) :- getIndividuoNotaAvaliacao(H, NA), getPairs(T, Tail).

ordenaPopulacao(Populacao, PopulacaoOrdenada):-
 getPairs(Populacao, Pairs),
 sort(1, @>=, Pairs, ParesOrdenados),
 pairs_values(ParesOrdenados, PopulacaoOrdenada).

% Avalia a populacao
avaliaPopulacao([], []).
avaliaPopulacao([H|T], [I|Tail]) :- avaliacaoIndividuo(H, I), avaliaPopulacao(T, Tail).

% Soma as avaliacoes dos individuos
somaAvaliacoes([], Soma, Soma).
somaAvaliacoes([H|T], Acc, Soma) :- getIndividuoNotaAvaliacao(H, NA), NewAcc is Acc + NA, somaAvaliacoes(T, NewAcc, Soma). 

% Seleciona pai para o crossover
selecionaPaiAux(Populacao, PaiIndice, ValorSorteado, Soma, I, Out) :-
 length(Populacao, TamanhoPopulacao), getByIndex(Populacao, I, Individuo), getIndividuoNotaAvaliacao(Individuo, NA),
 I < TamanhoPopulacao, Soma < ValorSorteado,
 NewPaiIndice is PaiIndice+1, NewSoma is Soma + NA, NewI is I+1,
 selecionaPaiAux(Populacao, NewPaiIndice, ValorSorteado, NewSoma, NewI, Out).
selecionaPaiAux(_, PaiIndice, _, _, _, PaiIndice).

selecionaPai(Populacao, SomaAvaliacao, PaiIndice) :- random(Rand), ValorSorteado is Rand * SomaAvaliacao,
 selecionaPaiAux(Populacao, -1, ValorSorteado, 0, 0, PaiIndice).

% Visualiza melhor individuo de uma geracao
visualizaGeracao(Populacao) :- ordenaPopulacao(Populacao, PO), getByIndex(PO, 0, I), visualizaIndividuo(I).

% Cria uma nova populacao
criaNovaPopulacao(_, 0, _, _, []).
criaNovaPopulacao(Populacao, Tamanho, SomaAvaliacao, TaxaMutacao, [NewF1, NewF2|Tail]) :-
 selecionaPai(Populacao, SomaAvaliacao, Pai1Indice),
 selecionaPai(Populacao, SomaAvaliacao, Pai2Indice),
 getByIndex(Populacao, Pai1Indice, P1), getByIndex(Populacao, Pai2Indice, P2),
 crossoverIndividuo(P1, P2, (F1, F2)),
 mutacaoIndividuo(F1, TaxaMutacao, NewF1), mutacaoIndividuo(F2, TaxaMutacao, NewF2),
 NewTamanho is Tamanho-1,
 criaNovaPopulacao(Populacao, NewTamanho, SomaAvaliacao, TaxaMutacao, Tail).

% Loop
loopAux(Populacao, SomaAvaliacao, TaxaMutacao, TamanhoPopulacao, (NPA, Melhor)) :-
 NewTamanho is TamanhoPopulacao / 2,
 criaNovaPopulacao(Populacao, NewTamanho, SomaAvaliacao, TaxaMutacao, NovaPopulacao),
 avaliaPopulacao(NovaPopulacao, NPA), ordenaPopulacao(NPA, PO), getByIndex(PO, 0, Melhor).

looping(_, _, _, 0, _, Melhor, _) :- write_ln("Melhor solucao: "), visualizaIndividuo(Melhor), 
 getIndividuoCromossomo(Melhor, C), getProdutos(P), getProdutosIndividuo(P, C, 0, PI), imprimeProdutos(PI).
looping(SomaAvaliacao, LimiteEspaco, TamanhoPopulacao, NumGeracoes, TaxaMutacao, MelhorIndividuo, Populacao) :-
 loopAux(Populacao, SomaAvaliacao, TaxaMutacao, TamanhoPopulacao, (NovaPopulacao, Melhor)),
 visualizaGeracao(NovaPopulacao), melhorIndividuo(MelhorIndividuo, Melhor, M),
 NewNumGeracoes is NumGeracoes-1, somaAvaliacoes(NovaPopulacao, 0, NewSomaAvaliacao),
 looping(NewSomaAvaliacao, LimiteEspaco, TamanhoPopulacao, NewNumGeracoes, TaxaMutacao, M, NovaPopulacao).

% Retorna os produtos escolhidos apartir do cromossomo do individuo
getProdutosIndividuo(_, [], _, []).
getProdutosIndividuo(Produtos, [1|T], Index, [P|Tail]) :-
 getByIndex(Produtos, Index, P), NewIndex is Index+1,
 getProdutosIndividuo(Produtos, T, NewIndex, Tail).
getProdutosIndividuo(Produtos, [_|T], Index, Out) :-
 NewIndex is Index+1, getProdutosIndividuo(Produtos, T, NewIndex, Out).

main :-
 write("Tamanho da populacao: "), write_ln(20),
 write("Limite: "), write_ln(3.0000),
 write("Taxa mutacao: "), write_ln(0.1),
 write("Geracao: "), write_ln(100),
 write_ln(""),

 getProdutos(Produtos),
 imprimeProdutos(Produtos),
 write_ln(""),
 getEspacos(Produtos, E),
 getValores(Produtos, V),
 
 initPopulacao(20, E, V, 3.0000, Populacao),
 ordenaPopulacao(Populacao, PO),
 avaliaPopulacao(Populacao, PA),
 somaAvaliacoes(PA, 0, SomaAvaliacao),
 visualizaGeracao(PA),
 getByIndex(PO, 0, Melhor),
 looping(SomaAvaliacao, 3.0000, 20, 100, 0.1, Melhor, PA).

File: genetic_algorithm/genetic_algorithm-haskell/genetic_algorithm.hs
{-# LANGUAGE OverloadedStrings, DeriveGeneric #-}

import Data.Aeson
import Control.Applicative ()
import qualified Data.ByteString.Lazy as B
import Network.HTTP.Conduit (simpleHttp)
import GHC.Generics ( Generic )
import Control.Monad ( MonadPlus(mzero) )
import System.IO.Unsafe ( unsafePerformIO )
import System.IO ()
import Control.Monad (mfilter)
import Numeric ( showFFloat )
import System.Random ( randomIO, randomRIO )
import Data.List ( sortBy )

data Produto = Produto {
    nome :: String,
    espaco :: Float,
    valor :: Float
} deriving (Show, Generic)

data Individuo = Individuo {
    espacos :: [Float],
    valores :: [Float],
    cromossomo :: [Int],
    espaco_usado :: Float,
    limite_espacos :: Float,
    geracao :: Int,
    nota_avaliacao :: Float
} deriving (Show) 

instance FromJSON Produto where
 parseJSON (Object v) =
    Produto <$> v .: "nome"
            <*> v .: "espaco"
            <*> v .: "valor"
 parseJSON _ = mzero

instance ToJSON Produto where
 toJSON (Produto nome espaco valor) =
    object [ "nome"   .= nome
           , "espaco" .= espaco
           , "valor"  .= valor ]

-- Retorna um intervalo de um array
sliceArray :: [Int] -> Int -> Int -> Int -> [Int]
sliceArray [] _ _ _ = []
sliceArray (h:t) x y cont
 | cont >= y = []
 | cont < x = sliceArray t x y (cont+1) 
 | otherwise = [h] ++ (sliceArray t x y (cont+1))
--

-- Concatena dois arrays
concatArray :: [Int] -> [Int] -> [Int]
concatArray a b = a ++ b

-- Converte um float para string
showFloat :: Float -> String
showFloat x = showFFloat Nothing x ""

-- Le um arquivo
read_File :: String -> IO B.ByteString
read_File path = B.readFile path

-- Extrai o arquivo json
readJSON :: Either String [t] -> [t]
readJSON entrada =
 case entrada of
  Left err -> []
  Right ps -> ps

-- Retorna a lista de produtos
getProdutos :: [Produto]
getProdutos = readJSON $ unsafePerformIO ( (eitherDecode <$> (read_File "../produtos.json")) :: IO (Either String [Produto]) )

-- Retorna o tamanho da maior string de nome, espaco e valor dos produtos 
getTamanhoStringProdutoComMaiorNome :: [Produto] -> Int -> Int
getTamanhoStringProdutoComMaiorNome [] aux = aux
getTamanhoStringProdutoComMaiorNome (h:t) aux 
 | length (nome h) > aux = getTamanhoStringProdutoComMaiorNome t (length (nome h))
 | otherwise = getTamanhoStringProdutoComMaiorNome t aux

getTamanhoStringProdutoComMaiorEspaco :: [Produto] -> Int -> Int
getTamanhoStringProdutoComMaiorEspaco [] aux = aux
getTamanhoStringProdutoComMaiorEspaco (h:t) aux 
 | length (showFloat (espaco h)) > aux = getTamanhoStringProdutoComMaiorEspaco t (length (showFloat (espaco h)))
 | otherwise = getTamanhoStringProdutoComMaiorEspaco t aux

getTamanhoStringProdutoComMaiorValor :: [Produto] -> Int -> Int
getTamanhoStringProdutoComMaiorValor [] aux = aux
getTamanhoStringProdutoComMaiorValor (h:t) aux 
 | length (showFloat (valor h)) > aux = getTamanhoStringProdutoComMaiorValor t (length (showFloat (valor h)))
 | otherwise = getTamanhoStringProdutoComMaiorValor t aux

getMaioresTamanhosStringProdutosInfo :: [Produto] -> [Int]
getMaioresTamanhosStringProdutosInfo p = [(getTamanhoStringProdutoComMaiorNome p 0)+1, (getTamanhoStringProdutoComMaiorEspaco p 0)+1, (getTamanhoStringProdutoComMaiorValor p 0)+1]
--

-- Imprime x espacos na mesma linha
imprimeEspacos :: Int -> IO()
imprimeEspacos 0 = putStr ""
imprimeEspacos x = do
  putStr " "
  imprimeEspacos (x-1)

-- Imprime os produtos
imprimeProdutosAux :: [Produto] -> [Int] -> IO()
imprimeProdutosAux [] _ = putStr ""
imprimeProdutosAux (h:t) [mn, me, mv] = do 
  putStr (nome h)
  imprimeEspacos (mn - (length (nome h)))
  putStr " - "

  putStr $ showFloat (valor h)
  imprimeEspacos (mv - (length $ showFloat (valor h)))
  putStr " - "

  putStr $ showFloat (espaco h)
  putStrLn ""

  imprimeProdutosAux t [mn, me, mv]

imprimeProdutos :: [Produto] -> IO()
imprimeProdutos produtos = do
  let t = getMaioresTamanhosStringProdutosInfo produtos 
  putStr "Produto"
  imprimeEspacos (t!!0 - 4)
  putStr "Valor"
  imprimeEspacos (t!!2 - 2)
  putStrLn "Espaco"
  imprimeProdutosAux produtos t
--

-- Gera um individuo
randomList :: Int -> IO([Int])
randomList 0 = return []
randomList n = do
  r  <- randomRIO (0,1)
  rs <- randomList (n-1) 
  return (r:rs)

initIndividuo :: [Float] -> [Float] -> Float -> Individuo
initIndividuo espacos valores limite = (Individuo espacos valores (unsafePerformIO (randomList (length espacos))) 0 limite 0 0)
--

-- Avalia um individuo
somaEspacoUsado :: [Float] -> [Int] -> Float
somaEspacoUsado [] [] = 0
somaEspacoUsado (e:t1) (c:t2)
 | c == 1 = e + (somaEspacoUsado t1 t2)
 | otherwise = somaEspacoUsado t1 t2

somaNotaAvaliacao :: [Float] -> [Int] -> Float
somaNotaAvaliacao [] [] = 0
somaNotaAvaliacao (e:t1) (c:t2)
 | c == 1 = e + (somaNotaAvaliacao t1 t2)
 | otherwise = somaNotaAvaliacao t1 t2

verificaSomaEspaco :: Float -> Float -> Float -> Float
verificaSomaEspaco soma_espaco limite nota
 | soma_espaco >= limite = 1.0
 | otherwise = nota

avaliacaoCromossomoIndividuo :: Individuo -> [Int] -> (Float, Float)
avaliacaoCromossomoIndividuo i cromossomo = do
 let soma_na = somaNotaAvaliacao (valores i) cromossomo
 let soma_espaco_usado = somaEspacoUsado (espacos i) cromossomo
 let soma_nota_avaliacao = verificaSomaEspaco soma_espaco_usado (limite_espacos i) soma_na
 (soma_nota_avaliacao, soma_espaco_usado)

avaliacaoIndividuo :: Individuo -> Individuo
avaliacaoIndividuo i = do
  let tuple = avaliacaoCromossomoIndividuo i (cromossomo i)
  (Individuo (espacos i) (valores i) (cromossomo i) (snd tuple) (limite_espacos i) (geracao i) (fst tuple))
--

-- Faz o crossover de invidivuos
corte :: Int -> Int
corte size 
 | x == 0 = corte size
 | otherwise = x 
 where x = unsafePerformIO(randomRIO (0, size))

crossoverIndividuo :: Individuo -> Individuo -> (Individuo, Individuo)
crossoverIndividuo a b = do
  let c = corte (length (cromossomo a) - 1)

  let filho1_pt1 = sliceArray (cromossomo b) 0 c 0 
  let filho1_pt2 = sliceArray (cromossomo a) c (length (cromossomo a)) 0
  let f1 = concatArray filho1_pt1 filho1_pt2

  let filho2_pt1 = sliceArray (cromossomo a) 0 c 0
  let filho2_pt2 = sliceArray (cromossomo b) c (length (cromossomo b)) 0
  let f2 = concatArray filho2_pt1 filho2_pt2

  let tuple1 = avaliacaoCromossomoIndividuo a f1
  let filho1 = Individuo (espacos a) (valores a) f1 (snd tuple1) (limite_espacos a) ((geracao a) + 1) (fst tuple1)

  let tuple2 = avaliacaoCromossomoIndividuo a f2
  let filho2 = Individuo (espacos a) (valores a) f1 (snd tuple2) (limite_espacos a) ((geracao a) + 1) (fst tuple2)
  
  (filho1, filho2)
--

-- Faz mutaçao em individuos
getCromossomoMutacao :: [Int] -> Float -> [Int]
getCromossomoMutacao [] _ = []
getCromossomoMutacao (h:t) taxa
 | (rand < taxa) && (h == 1) = [0] ++ (getCromossomoMutacao t taxa)
 | (rand < taxa) && (h == 0) = [1] ++ (getCromossomoMutacao t taxa)
 | otherwise = [h] ++ (getCromossomoMutacao t taxa)
 where rand = unsafePerformIO(randomIO)

mutacaoIndividuo :: Individuo -> Float -> Individuo
mutacaoIndividuo i taxa = do
  let novo_cromossomo = getCromossomoMutacao (cromossomo i) taxa
  (Individuo (espacos i) (valores i) novo_cromossomo (espaco_usado i) (limite_espacos i) (geracao i) (nota_avaliacao i))
--

-- Cria a populacao
initPopulacao :: Float -> [Float] -> [Float] -> Float -> [Individuo]
initPopulacao 0 _ _ _ = []
initPopulacao tamanho espacos valores limite_espacos = [initIndividuo espacos valores limite_espacos] ++ initPopulacao (tamanho-1) espacos valores limite_espacos

-- Retorna o melhor individuo
melhorIndividuo :: Individuo -> Individuo -> Individuo
melhorIndividuo a b 
 | (nota_avaliacao a) >= (nota_avaliacao b) = a 
 | otherwise = b

-- Ordena populacao
ordenaPopulacaoAux :: Individuo -> Individuo -> Ordering
ordenaPopulacaoAux i1 i2
 | (nota_avaliacao i1) > (nota_avaliacao i2) = GT
 | otherwise  = LT

ordenaPopulacao :: [Individuo] -> [Individuo]
ordenaPopulacao populacao = reverse (sortBy ordenaPopulacaoAux populacao)
--

-- Avalia populacao
avaliaPopulacao :: [Individuo] -> [Individuo]
avaliaPopulacao [] = []
avaliaPopulacao (h:t) = [avaliacaoIndividuo h] ++ (avaliaPopulacao t)

-- Soma avaliacoes dos individuos da populacao
somaAvaliacoes :: [Individuo] -> Float
somaAvaliacoes [] = 0
somaAvaliacoes (h:t) = (nota_avaliacao h) + (somaAvaliacoes t) 

-- Seleciona pai para o crossover 
selecionaPaiAux :: [Individuo] -> Int -> Float -> Float -> Int -> Int
selecionaPaiAux populacao pai_indice valor_sorteado soma i 
 | i < (length populacao) && soma < valor_sorteado = selecionaPaiAux populacao (pai_indice+1) valor_sorteado (soma + (nota_avaliacao (populacao!!i))) (i+1)
 | otherwise = pai_indice

selecionaPai :: [Individuo] -> Float -> Int
selecionaPai populacao soma_avaliacao = do
  let valor_sorteado = unsafePerformIO(randomIO) * soma_avaliacao
  (selecionaPaiAux populacao (-1) valor_sorteado 0 0)
--

-- Gera a visualizacao de um individuo
visualizaIndividuo :: Individuo -> IO()
visualizaIndividuo i = do
  putStr "Geracao: "
  print (geracao i)
  putStr "Nota avaliacao R$: "
  print (nota_avaliacao i)
  putStr "Espaco usado: "
  print (espaco_usado i)
  putStr "Cromossomo: " 
  print (cromossomo i)
  putStrLn ""

-- Gera a visualizacao de uma geracao
visualizaGeracao :: [Individuo] -> IO()
visualizaGeracao [] = putStr ""
visualizaGeracao populacao = visualizaIndividuo $ head populacao

-- Loop de geracoes
criaNovaPopulacao :: [Individuo] -> Int -> Float -> Float -> [Individuo]
criaNovaPopulacao populacao tamanho soma taxa
 | tamanho == 0 = []
 | otherwise = [f1, f2] ++ criaNovaPopulacao populacao (tamanho-1) soma taxa
 where pai1 = selecionaPai populacao soma
       pai2 = selecionaPai populacao soma
       filhos = crossoverIndividuo (populacao!!pai1) (populacao!!pai2) 
       f1 = mutacaoIndividuo (fst filhos) taxa
       f2 = mutacaoIndividuo (snd filhos) taxa

loopAux :: [Individuo] -> Float -> Float -> Float -> Individuo -> ([Individuo], Individuo)
loopAux populacao soma taxa tamanho melhor = do
  let t = floor $ tamanho / 2.0
  let nova_populacao = criaNovaPopulacao populacao t soma taxa
  let nova_populacao_avaliada = avaliaPopulacao nova_populacao
  let nova_populacao_ordenada = ordenaPopulacao nova_populacao_avaliada
  let melhor_individuo = head nova_populacao_ordenada
  (nova_populacao_avaliada, melhor_individuo)

looping :: Float -> Float -> Float -> Int -> Float -> Individuo -> [Individuo] -> [Individuo] 
looping _ _ _ 0 _ melhor_individuo _ = [melhor_individuo]
looping soma limite tamanho num_geracoes taxa melhor_individuo populacao = do
  let tuple = loopAux populacao soma taxa tamanho melhor_individuo
  (fst tuple) ++ (looping (somaAvaliacoes (fst tuple)) limite tamanho (num_geracoes-1) taxa (melhorIndividuo (snd tuple) melhor_individuo) (fst tuple))
--

-- Aux
print_evolucao :: [Individuo] -> IO()
print_evolucao [i] = do 
  putStrLn "Melhor solucao: "
  visualizaIndividuo i
print_evolucao (h:t) = do
  visualizaIndividuo h
  print_evolucao t

getEspacos :: [Produto] -> [Float]
getEspacos [] = []
getEspacos (h:t) = [(espaco h)] ++ (getEspacos t)

getValores :: [Produto] -> [Float]
getValores [] = []
getValores (h:t) = [(valor h)] ++ (getValores t)

getNomes :: [Produto] -> [String]
getNomes [] = []
getNomes (h:t) = [(nome h)] ++ (getNomes t)
--

getProdutosIndividuo :: [Produto] -> [Int] -> Int -> [Produto]
getProdutosIndividuo _ [] _ = []
getProdutosIndividuo produtos (1:t) index = [produtos!!index] ++ (getProdutosIndividuo produtos t (index+1))
getProdutosIndividuo produtos (_:t) index = getProdutosIndividuo produtos t (index+1)

-- Para instalar pacotes: cabal v1-install {pacote}
-- Para compilar: runhaskell genetic_algorithm.hs
main :: IO()
main = do 
  let tamanhoPopulacao = 20
  let limite = 3.0000
  let taxaMutacao = 0.1
  let geracoes = 100

  putStrLn ""
  putStr "Tamanho da populacao: "
  print tamanhoPopulacao
  putStr "Limite: "
  print limite
  putStr "Taxa mutacao: "
  putStrLn $ showFloat taxaMutacao
  putStr "Geracao: "
  print geracoes
  putStrLn ""

  let p = getProdutos
  imprimeProdutos p
  putStrLn ""

  let espacos = getEspacos p 
  let valores = getValores p 
  let nomes = getNomes p 

  let populacao = initPopulacao tamanhoPopulacao espacos valores limite
  let populacao_avaliada = avaliaPopulacao populacao
  let populacao_ordenada = ordenaPopulacao populacao_avaliada
  let melhor_individuo = head populacao_ordenada
  visualizaGeracao populacao_ordenada
  
  let evolucao = looping (somaAvaliacoes populacao_ordenada) limite tamanhoPopulacao geracoes taxaMutacao melhor_individuo populacao_ordenada
  print_evolucao evolucao
  let pi = getProdutosIndividuo p (cromossomo (evolucao!!((length evolucao)-1))) 0
  imprimeProdutos pi